feat: 회원가입 및 로그인 기능 구현

sksqlrlek opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**

**TO-BE**
1. 회원(user) 및 판매자(seller) 회원가입 및 로그인 기능 구현
2. mailgun을 이용한 이메일 인증(테스트 미완)
3. jwt를 이용한 암호화

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- [ ] API 테스트

sksqlrlek committed:
회원가입 및 로그인 기능 구현

1. 회원(user) 및 판매자(seller) 회원가입 및 로그인 기능 구현
2. mailgun을 이용한 이메일 인증(테스트 x)
3. jwt를 이용한 암호화

sksqlrlek committed:
feat: 회원가입 및 로그인 기능 구현

- 회원(user) 및 판매자(seller) 회원가입 및 로그인 구현
- mailgun을 이용한 이메일 인증 기능 추가 (테스트 미완)
- jwt를 이용한 토큰 암호화 적용

mingming-mentor commented on Game-domain/src/main/java/com/user/gamedomain/config/JwtAuthenticationProvider.java:
```diff
@@ -0,0 +1,49 @@
+
+public class JwtAuthenticationProvider {
+
+  private String secretKey = "secretKey";
```
이 부분은 추후 변경 예정이실까요? secretKey를 주입받던가 해야할 것 같아서요.

mingming-mentor commented on Game-domain/src/main/java/com/user/gamedomain/config/JwtAuthenticationProvider.java:
```diff
@@ -0,0 +1,49 @@
+        .compact();
+  }
+
+  public boolean validateToken(String jwtToken) {
```
validate라는 이름은 보통 성공시 void, 에러시 exception을 던지는 메소드에 붙이는 이름입니다. 
현재 작성하신 메소드의 경우 isValid 등의 이름이 좀더 적합합니다. isXX라는 메소드는 boolean을 반환하는 메소드에 붙이는 네이밍입니다.

mingming-mentor commented on Game-domain/src/main/java/com/user/gamedomain/util/Aes256Util.java:
```diff
@@ -0,0 +1,45 @@
+public class Aes256Util {
+
+  public static String alg = "AES/CBC/PKCS5Padding";
+  private static final String KEY = "GameSell2025AES256SecureKeyTest1";
```
secret을 이용하여 민감한 값들은 관리를 하시면 좋을 것 같습니다
- https://velog.io/@sun1203/Github-Action-Spring-Application.properties-%EB%AF%BC%EA%B0%90%EC%A0%95%EB%B3%B4-%EA%B4%80%EB%A6%AC%ED%95%98%EA%B8%B0
- https://velog.io/@seokjun0915/GitHub-Action%EC%97%90%EC%84%9C-Spring-Boot-%ED%99%98%EA%B2%BD%EB%B3%80%EC%88%98-%EC%84%A4%EC%A0%95%ED%95%98%EA%B8%B0

mingming-mentor commented on Game-domain/src/main/java/com/user/gamedomain/util/Aes256Util.java:
```diff
@@ -0,0 +1,45 @@
+      byte[] encrypted = cipher.doFinal(text.getBytes(StandardCharsets.UTF_8));
+      return Base64.encodeBase64String(encrypted);
+    } catch (Exception e) {
+      e.printStackTrace();
```
log를 찍으시면 좋겠습니다. 
https://tmdrnr96.tistory.com/43

mingming-mentor commented on Game-domain/src/main/java/com/user/gamedomain/util/Aes256Util.java:
```diff
@@ -0,0 +1,45 @@
+      return Base64.encodeBase64String(encrypted);
+    } catch (Exception e) {
+      e.printStackTrace();
+      return null;
```
exception을 잡고 null을 return하도록 되어있는데요, exception 자체를 던질 필요는 없나요?

mingming-mentor replied:
암호화가 제대로 안되었을 텐데, 이를 null로 임의 처리해도 되는지 궁금합니다. 
exception을 던져 동작자체가 안되도록 해야하지 않나 싶네요.

sksqlrlek replied:
@mingming-mentor 
말씀해주신대로 log를 넣고, Exception을 던지는 코드를 짜봤는데 테스트 해본 결과를 보니 암호화, 복호화 과정에서 오류가 있는 것 같습니다.
처음 코드를 짰을 때 이 부분을 제대로 테스트를 못 한 것 같네요.
현재 무엇이 문제인지 잘 모르겠습니다.

mingming-mentor replied:
@sksqlrlek 님 지금 말씀해주시는 것만으로는 제가 에러 원인을 같이 찾아봐드리기 어렵습니다. 
어떤 에러가 떴는지, debug과정에서는 어땠는지를 같이 첨부해서 알려주세요. 화면첨부해서 올려주셔도 괜찮습니다

sksqlrlek replied:
@mingming-mentor 
밍밍멘토님, 다시 테스트 해보니 문제 없이 잘 실행됩니다. 번거롭게 해드려 죄송합니다!

mingming-mentor commented on user-api/src/main/java/com/zerobase/GameSell/user/service/SignUpUserService.java:
```diff
@@ -0,0 +1,61 @@
+      return user.getVerifyExpiredAt();
+    }
+    //todo : 예외 처리
+    throw new UserException(NOT_FOUND_USER);
```
Optional로 된 로컬 변수를 만드는 것은 지양되면 좋겠습니다
```suggestion
    User user = userRepository.findById(userId).orElseThrow(() -> new UserException(NOT_FOUND_USER));

    user.setVerificationCode(verificationCode);
    user.setVerifyExpiredAt(LocalDateTime.now().plusDays(1));
    return user.getVerifyExpiredAt();
```

mingming-mentor commented on user-api/src/main/java/com/zerobase/GameSell/user/service/SignUpUserService.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Transactional
+  public LocalDateTime changeUserValidateEmail(Long userId, String verificationCode) {
```
`changeUserVerifyExpiredDateTime`이 좀더 적합한 네이밍일 것 같습니다. 
현재 네이밍으로는 유저의 이메일을 변경하겠다는 의미를 담고 있네요

mingming-mentor commented on user-api/src/main/java/com/zerobase/GameSell/user/domain/model/User.java:
```diff
@@ -0,0 +1,60 @@
+
+  private LocalDateTime verifyExpiredAt;
+  private String verificationCode;
+  private boolean verify;
```
ERD 에 반영해주시면 좋을 것 같습니다. 
또한 verify보단 verifiedDateTime을 넣어서 인증된 날짜와 인증 여부를 함께 저장하는 것이 좀더 나을 것 같습니다.
하나의 컬럼으로 여러개의 정보를 보여줄 수 있으니깐요

mingming-mentor commented on user-api/src/main/java/com/zerobase/GameSell/user/domain/model/User.java:
```diff
@@ -0,0 +1,60 @@
+  @Column(nullable = false)
+  private int age;
+  @Column(nullable = false, length = 1)
+  private String gender;
```
enum으로 관리해도 좋을 것 같네요

mingming-mentor commented on user-api/src/main/java/com/zerobase/GameSell/user/domain/model/User.java:
```diff
@@ -0,0 +1,60 @@
+  @Column(nullable = false, length = 50)
+  private String password;
+  @Column(nullable = false)
+  private int age;
```
age보단 birthDate가 좀더 낫겠습니다. age라면 매년 해가 바뀔때마다 모든 유저에 대해서 업데이트를 해줘야합니다.

mingming-mentor commented on user-api/src/main/java/com/zerobase/GameSell/user/service/SignUpUserService.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  public boolean isEmailExist(String email) {
+    return userRepository.findByEmail(email.toLowerCase(Locale.ROOT))
```
다른 findByEmail을 체크하는 부분에서는 이런 처리가 없어보이는데요. (UserService의 findByIdAndEmail 등)
save시에 toLowerCase를 없애도 될것 같은데, 넣은 이유가 있나요?

sksqlrlek replied:
보통 이메일의 경우 대소문자 구분 없이 사용되는 것을 생각해 처음 저장할 때, toLowerCase를 이용해 소문자로 저장하고
나중에 이메일 관련해서 작업할 때도 소문자로 찾을 수 있게 하는 것이 맞다고 생각하여 입력했습니다.
그래서 findByEmail에도 추가했습니다.

mingming-mentor reviewed (approved):
수고 많으셨습니다

sksqlrlek committed:
fix, refactor, chore: 회원가입 및 로그인 기능 개선

1. 오타 수정
2. 메소드 이름 수정
3. 민감한 값을 gitgub action을 이용해 암호화 및 주입
4. print 문을 log 문으로 수정
5. ERD 수정

sksqlrlek committed:
wip: 병합 전 임시 커밋

sksqlrlek committed:
Merge branch 'email-verification' into feature/Signup

# Conflicts:
#	Game-domain/build.gradle
#	Game-domain/src/main/java/com/user/gamedomain/config/JwtAuthenticationProvider.java
#	Game-domain/src/main/java/com/user/gamedomain/util/Aes256Util.java
#	user-api/build.gradle
#	user-api/data/h2db.mv.db
#	user-api/src/main/java/com/zerobase/GameSell/user/UserApplication.java
#	user-api/src/main/java/com/zerobase/GameSell/user/application/SignUpApplication.java
#	user-api/src/main/java/com/zerobase/GameSell/user/config/JwtConfig.java
#	user-api/src/main/java/com/zerobase/GameSell/user/controller/SignUpController.java
#	user-api/src/main/java/com/zerobase/GameSell/user/domain/SignUpForm.java
#	user-api/src/main/java/com/zerobase/GameSell/user/domain/model/User.java
#	user-api/src/main/java/com/zerobase/GameSell/user/service/SignUpUserService.java
#	user-api/src/main/java/com/zerobase/GameSell/user/service/UserService.java
#	user-api/src/main/resources/application.properties
#	user-api/src/test/resources/application.properties
